lib/tesla/adapter/finch.ex

```diff
@@ -69,20 +74,74 @@ if Code.ensure_loaded?(Finch) do
       end
     end
 
-    defp request(name, method, url, headers, %Multipart{} = mp, opts) do
+    defp build(method, url, headers, %Multipart{} = mp) do
       headers = headers ++ Multipart.headers(mp)
       body = Multipart.body(mp) |> Enum.to_list()
 
-      request(name, method, url, headers, body, opts)
+      build(method, url, headers, body)
+    end
+
+    defp build(method, url, headers, %Stream{} = body_stream) do
+      build(method, url, headers, {:stream, body_stream})
     end
 
-    defp request(_name, _method, _url, _headers, %Stream{}, _opts) do
-      raise "Streaming is not supported by this adapter!"
+    defp build(method, url, headers, body_stream_fun) when is_function(body_stream_fun) do
+      build(method, url, headers, {:stream, body_stream_fun})
     end
 
-    defp request(name, method, url, headers, body, opts) do
+    defp build(method, url, headers, body) do
       Finch.build(method, url, headers, body)
-      |> Finch.request(name, opts)
+    end
+
+    defp request(req, name, req_opts, opts) do
+      case opts[:response] do
+        :stream -> stream(req, name, req_opts)
+        nil -> Finch.request(req, name, req_opts)
+        other -> raise "Unknown response option: #{inspect(other)}"
+      end
+    end
+
+    defp stream(req, name, opts) do
+      owner = self()
+      ref = make_ref()
+
+      fun = fn
+        {:status, status}, _acc -> status
+        {:headers, headers}, status -> send(owner, {ref, {:status, status, headers}})
+        {:data, data}, _acc -> send(owner, {ref, {:data, data}})
+      end
+
+      task =
+        Task.async(fn ->
+          case Finch.stream(req, name, nil, fun, opts) do
+            {:ok, _acc} -> send(owner, {ref, :eof})
+            {:error, error} -> send(owner, {ref, {:error, error}})
+          end
+        end)
```

Comment:
I was curious about the need to spawn a process. I tried a version of this without the `Task.async`. The existing process seemed to handle the send and receive without wrapping `Finch.stream` in `Task.async`. It returned a stream function as the body of my response. I was able to use that to stream a CSV file and then parse it using NimbleCSV's stream functions. I may be missing something about the way that Tesla behaves, though. Is there a danger of the existing process ending before the stream has been fully consumed? Is that it is being wrapped in `Task.async`?

Here is a sample of the variation that I tried.

```elixir
defmodule MyStream do

  def stream(url) do
    request = build_request(url)
    owner = self()
    ref = make_ref()

    fun = fn
      {:status, value}, _status ->
        value

      {:headers, headers}, status ->
        send(owner, {ref, {:status, status, headers}})
        status

      {:data, data}, status ->
        send(owner, {ref, {:data, data}})
        status
    end

    case Finch.stream(request, MyFinch, nil, fun) do
      {:ok, _status} ->
        send(owner, {ref, :eof})

      {:error, error} ->
        send(owner, {ref, {:error, error}})
    end

    receive do
      {^ref, {:status, status, headers}} ->
        body =
          Stream.unfold(nil, fn _ ->
            receive do
              {^ref, {:data, data}} ->
                {data, nil}

              {^ref, :eof} ->
                nil
            after
              1_500 ->
                nil
            end
          end)

        {:ok, %Finch.Response{status: status, headers: headers, body: body}}
    after
      1_500 ->
        {:error, :timeout}
    end
  end

  defp build_request(url) do
    Finch.build(:get, url)
  end
end
```

These are the sorts of commands I ran in iex to check that the body returned was a stream. I also added logging for the different receive blocks and was able to see that we received the `{^ref, :eof}` message after consuming the full stream.

```elixir
alias NimbleCSV.RFC4180, as: CSV
iex(18)> url = "some-url-to-a-csv"
"some-url-to-a-csv"
iex(18)> {:ok, %{body: body}} = MyStream.stream(url)
{:ok,
 %Finch.Response{
   status: 200,
   body: #Function<63.124013645/2 in Stream.unfold/2>,
   headers: [
     {"content-length", "9530"},
     {"content-type", "text/csv"}
   ]
 }}
iex(19)> line_stream = CSV.to_line_stream(body)
#Function<60.124013645/2 in Stream.transform/3>
iex(20)> parsed = CSV.parse_stream(line_stream, skip_headers: false)
#Function<61.124013645/2 in Stream.transform/4>
iex(21)> as_list = Enum.to_list(parsed)
...
```

Reply:
Note that `Finch.stream` is blocking - it will not return untill the stream has finished. It works in console because after `Finch.stream` returns you can read all messages from the mailbox, but if you introduce latency in steam data you can observe that current version returns early (after getting status & headers) and the one without Task will block untill the whole stream is read.